### PR TITLE
Allow to define where to save Action results attributes

### PR DESCRIPTION
**Pull request by nicolocaruso85**

### CODE REVIEW

**Comment by nicolocaruso85** [comment]
fixes https://github.com/opendialogai/core/issues/203

**Comment by stu-greenshoots** [comment]
@istos - would be good to look through this one together

File: src/Conversation/Intent.php [hunk 1]
```diff
@@ -275,6 +304,31 @@ public function getExpectedAttributeContexts(): Map
...
+        try {
+            /** @var ExpectedAttribute $expectedActionAttribute */
+            foreach ($this->getExpectedActionAttributes() as $expectedActionAttribute) {
+                $attributesActionContexts->put(
+                    ContextParser::determineAttributeId($expectedActionAttribute->getId()),
+                    ContextParser::determineContextId($expectedActionAttribute->getId())
+                );
+            }
+        } catch (NodeDoesNotExistException $e) {
+            // nothing
```
Comment (stu-greenshoots) [inline src/Conversation/Intent.php]: Can we log a warning level message here at least

File: src/ConversationEngine/ConversationEngine.php [hunk 2]
```diff
@@ -492,13 +497,57 @@ public function performIntentAction(UserContext $userContext, Intent $nextIntent
...
+     * Stores the attributes from an Action to a context.
+     * Expected action attributes are retrieved from the Intent to determine which context each
+     * attribute should be saved to.
+     * If one is not defined for the attribute, it is saved to the user context
+     *
+     * @param ActionResult $actionResult
+     * @param UserContext $userContext
+     * @param Map $expectedActionAttributes
+     */
+    private function storeActionResult(ActionResult $actionResult, UserContext $userContext, Map $expectedActionAttributes)
```
Comment (stu-greenshoots) [inline src/ConversationEngine/ConversationEngine.php]: Can we refactor this to share code with saving intent attributes?

File: src/ConversationEngine/ConversationEngine.php [hunk 3]
```diff
@@ -492,13 +497,57 @@ public function performIntentAction(UserContext $userContext, Intent $nextIntent
...
+                        Log::error(sprintf('Attribute context %s does not exist, using user context', $contextId));
+                    }
+                }
+            }
+
+            $context->addAttribute($attribute);
+        }
+
+        // TODO - is there a better way of doing this? Each context could have it's own tear down method to deal with persisting
+        if ($userContextUpdated) {
```
Comment (stu-greenshoots) [inline src/ConversationEngine/ConversationEngine.php]: Can this work like with intents and save all contexts

**Comment by istos** [comment]
this addressses #134

**Comment by stu-greenshoots** [comment]
@nicolocaruso85 Can you make a change to the YAML so that it looks like this:

```
action:
              id: action.core.example
             input_attributes:
               - user.name 
             output_attributes:
                - user.first_name
                - session.full_name
```


Here, `output_attributes` would work like `expected attributes` and input attributes would select which context to pull the attributes that get pulled into the ActionInput here https://github.com/opendialogai/core/blob/develop/src/ActionEngine/Service/ActionEngine.php#L103

This would require the conversation engine to inform the action engine of the `input_attributes`

File: src/ConversationEngine/ConversationStore/EIModels/EIModelIntent.php [hunk 4]
```diff
@@ -396,6 +465,30 @@ public function getExpectedAttributeContexts()
         return $attributesContexts;
     }
 
+    public function getInputActionAttributeContexts()
```
Comment (stu-greenshoots) [inline src/ConversationEngine/ConversationStore/EIModels/EIModelIntent.php]: Can you add the return type and phpdoc

File: src/ConversationEngine/ConversationStore/EIModels/EIModelIntent.php [hunk 5]
```diff
@@ -396,6 +465,30 @@ public function getExpectedAttributeContexts()
...
+        foreach ($this->inputActionAttributes as $inputActionAttribute) {
+            $attributesActionContexts->put(
+                ContextParser::determineAttributeId($inputActionAttribute),
+                ContextParser::determineContextId($inputActionAttribute)
+            );
+        }
+        return $attributesActionContexts;
+    }
+
+    public function getOutputActionAttributeContexts()
```
Comment (stu-greenshoots) [inline src/ConversationEngine/ConversationStore/EIModels/EIModelIntent.php]: Can you add the return type and phpdoc

**Comment by stu-greenshoots** [comment]
This looks good - happy with the changes here.

1 thing -can you add some tests for conversations that do not pass in the input attributes to show what happens

(I have raised an issue in core to update the OD docs around this as actions in conversation flow now need to define their input attributes)

**Pull request merged**